pkg/actions/addon/create.go

```diff
@@ -221,18 +218,19 @@ func (a *Manager) createRoleUsingAttachPolicyARNs(addon *api.Addon, namespace, s
 	return resourceSet.OutputRole, nil
 }
 
-func (a *Manager) createRoleUsingAttachPolicy(addon *api.Addon, namespace, serviceAccount string) (string, error) {
-	resourceSet := builder.NewIAMRoleResourceSetWithAttachPolicy(addon.Name, namespace, serviceAccount, addon.PermissionsBoundary, addon.AttachPolicy, a.oidcManager)
-	err := resourceSet.AddAllResources()
-	if err != nil {
-		return "", err
-	}
-
-	err = a.createStack(resourceSet, addon)
-	if err != nil {
-		return "", err
+func (a *Manager) createRoleResourceSet(addon *api.Addon, namespace, serviceAccount string) (*builder.IAMRoleResourceSet, error) {
+	var resourceSet *builder.IAMRoleResourceSet
+	if len(addon.AttachPolicyARNs) != 0 {
+		logger.Info("creating role using provided policies ARNs")
+		resourceSet = builder.NewIAMRoleResourceSetWithAttachPolicyARNs(addon.Name, namespace, serviceAccount, addon.PermissionsBoundary, addon.AttachPolicyARNs, a.oidcManager)
+	} else if addon.WellKnownPolicies.HasPolicy() {
+		logger.Info("creating role using provided well known policies")
+		resourceSet = builder.NewIAMRoleResourceSetWithWellKnownPolicies(addon.Name, namespace, serviceAccount, addon.PermissionsBoundary, addon.WellKnownPolicies, a.oidcManager)
+	} else {
+		logger.Info("creating role using provided policies")
+		resourceSet = builder.NewIAMRoleResourceSetWithAttachPolicy(addon.Name, namespace, serviceAccount, addon.PermissionsBoundary, addon.AttachPolicy, a.oidcManager)
 	}
-	return resourceSet.OutputRole, nil
+	return resourceSet, resourceSet.AddAllResources()
```

Comment:
This is so weird. It returns the resource, but then also returns the output of calling AddAllResources on the created resource. The cognitive load of this "simple" function is pretty high. I would prefer this to be separated and the caller has the responsibility to further add all resources if necessary. The `createroleResourceSet` doesn't suggest that this will also add all of them. It suggest construction.

Reply:
I think whats weird is how the resourceSets work. You use a constructor to "build" it, but its only actually built/useful after you've called `AddAllResources`. I could move this call further up, but its just moving the error handling around even more :shrug:

Reply:
Ah, well that's annoying. :D Fine, leave it as is :)